File: Trees/572. 另一棵树的子树/572. 另一棵树的子树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: TreeNode, subRoot: TreeNode) -> bool:
        """
        :type s: TreeNode
        :type t: TreeNode
        :rtype: bool
        """
        if not root and not subRoot : return True
        if not root or not subRoot : return False
        return self.isSameTree(root, subRoot) or self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)
        
    def isSameTree(self, s, t):
        if not s and not t : return True
        if not s or not t : return False
        return s.val == t.val and self.isSameTree(s.left, t.left) and self.isSameTree(s.right, t.right)

File: Trees/572. 另一棵树的子树.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isSubtree(TreeNode* root, TreeNode* subRoot) {
        // 都不等於nullptr
        if(root == nullptr && subRoot == nullptr) return true;
        // 僅其中一個等於nullptr
        if(root == nullptr || subRoot == nullptr) return false;

        return isSameTree(root, subRoot) or isSubtree(root->left, subRoot) or isSubtree(root->right, subRoot);
    }

private:
    bool isSameTree(TreeNode* p, TreeNode* q) {
        // 都不等於nullptr
        if(p == nullptr && q == nullptr) return true;
        // 僅其中一個等於nullptr
        if(p == nullptr || q == nullptr) return false;

        return p->val == q->val and isSameTree(p->left, q->left) and isSameTree(p->right, q->right);
    }
};